Caching stuff

scjackson opened the pull request:
@jordangarcia @mikeng13 

Added a number of options/functionality around caching including:

1. `getCacheValues` and `clearCacheValues` to `Reactor`. This will allow users to monitor their own cache usage and clear it if need be.
2. `useCache` and `maxItemsToCache` as config options for `Reactor`. This will allow users to disable caching completely, or limit the number of items that are stored in cache. Items are evicted from cache on a standard LRU policy.
3. A utility function for getters, `addOptionsToGetter`, which accepts a getter and augments it with options. Currently supported options are `useCache`, which allows users to specify whether the getter value should be cached on evaluate, and `cacheKey`, which allows users to specify their own cache key.

loganlinn commented on src/getter.js:
```diff
@@ -102,6 +102,35 @@ function getStoreDeps(getter) {
+ * @param {*} options.cacheKey
+ * @returns {getter}
+ */
+function addGetterOptions(getter, options) {
```
This API change to introduce the concept of "getter options" is a little awkward imo.

Currently, it will replace any previously set options which could lead to unexpected behavior given its name.
An alternative would be to call this `setGetterOptions` to indicate it would replace, not patch, the options.
But I feel like the API is being conflated with implementation details of hidden `__options` object. The alternative that I would suggest is two separate methods to configure these options.

cc @jordangarcia since this is API design considerations

loganlinn replied:
@jordangarcia  Another thing to consider is that it's probably natural from an API standpoint that getters are configured where they are defined. It may be time to introduce a formal `Getter()` constructor... :)

jordangarcia replied:
I would like to see a `Getter` constructor.

```js
var getter = Getter(['items']);
// or
var getter = Getter(['items'], {
  useCache: false
});
```

scjackson replied:
@jordangarcia Is there an advantage to using a constructor outside of style preference?

loganlinn commented on src/getter.js:
```diff
@@ -102,6 +102,35 @@ function getStoreDeps(getter) {
+ */
+function addGetterOptions(getter, options) {
+  if (!isKeyPath(getter) && !isGetter(getter)) {
+    throw new Error('createGetter must be passed a keyPath or Getter')
```
Looks like the error string was copied from `createGetter()` and should be updated

loganlinn commented on src/getter.js:
```diff
@@ -102,6 +102,35 @@ function getStoreDeps(getter) {
+    throw new Error('createGetter must be passed a keyPath or Getter')
+  }
+
+  if (getter.hasOwnProperty('__options')) {
```
Storing these options on `getter` could lead to surprising behavior if `getter` a key path that should be carefully considered and definitely documented. 

A newcomer to nuclear-js might think this would change the cache key (globally):

```js
addGetterOptions(['items'], {cacheKey: "items_cache"});
reactor.evaluate(['items']);
```

jordangarcia replied:
I think a `Getter` constructor alleviates some of this confusion, as it's clear that you are creating an instance of something and the options you provide with it would be tied to that instance.

loganlinn commented:
Rather than exposing crude controls to internal cache and caching behavior, it seems to me a better design would be to introduce a caching interface abstraction to allow the user to customize at will. This would avoid adding complexity to nuclear-js to permit additional caching algorithms (ie `maxItemsToCache`) and controls (ie `clearCacheValues`) for this particular use case.

jordangarcia commented on src/reactor.js:
```diff
@@ -285,6 +290,21 @@ class Reactor {
+   * Clear all cached values
+   */
+  clearCacheValues() {
+    this.reactorState = this.reactorState.set('cache', Immutable.Map())
```
what is the use case of this api method?

scjackson replied:
added as an afterthought, thought it might be useful in some cases of switching views in a spa. will remove

scjackson commented:
@loganlinn A cache abstraction would be nice. Did you have something in mind?

scjackson commented:
Before I make any changes, I want to make sure I point out the driving motivator of this change, which is that caching as it is currently implemented is broken. There is no cache cleanup, and as such memory will grow linearly over time. This is a real problem for single page apps and data intensive pages. IMO we need to fix or mitigate this out of the box for users. 

I'd like to address this in one of two ways:

1. Stick with what I've done above and add a default for `maxItemsToCache`. This would provide at least some ceiling to memory growth. I'm fine with removing some of the extra surface area I added.
2. Turn caching off by default and remove the LRU stuff.

In either case, adding a cache abstraction would be a logical next iteration to help power users fine tune their apps, but @loganlinn I think we need to address the base case first.

If you have better ideas, or a good handle on what a caching interface would look like, please let me know.

jordangarcia commented:
@scjackson I agree that the lack of cache invalidation / cleanup is a core problem that needs to be addressed.

I believe any viable candidate that ships with NuclearJS must (1) prevent the infinitely linear growth of cached data over time and (2) maintain the caching performance benefits that already exist in NuclearJS.

Here is what I propose, when a Getter is unobserved by invoking the returned unwatchFn, if no other observers are using that getter then we remove any cache entries for it.  This provides a more natural cache GC when a Getter is no longer "hot".

Thoughts?

scjackson commented:
@jordangarcia I like that. Would you be ok with disabling caching for unobserved getters (ie things that are just `flux.evaluate`d)

jordangarcia commented:
Hm, good point.  I am a little hesitant to change a lot of the out of the box caching behavior as it may have performance impacts on people using it.  Imagine a computationally expensive `evalaute` with a small memory footprint.   

I think having the option on a Getter to `alwaysCache` is reasonable.  `Getter(['items'], { alwaysCache: true })`

The more I think about it, the more I am leaning towards a hybrid approach, where we use intelligent unobserve logic to do cache cleanup and have a pretty high maxItems limit that serves as a cache ceiling for things like `flux.evaluate` cached Getters.

scjackson commented:
@jordangarcia So this would include:

1. `alwaysCache`, `skipCache`, and `cacheKey` options for getters, specified via `Getter` constructor
2. Cached values for observed getters are dropped on unobserve (this would help us a lot)
3. Add a default `maxItems` limit -- I propose 1000
4. (Optional) `useCache` option at the reactor level. Might not be as relevant with the steps taken above, but could be useful for those more concerned with memory usage than speed.

jordangarcia commented:
So if we change the default caching behavior of `reactor.evaluate` to not cache unless `alwaysCache` is true and we GC cache on Getter unobserve, is there ever a need for `skipCache` entirely?

The only use case I can think of is if you dont want the getter ever getting into cache during `observe`, which there may be a case where you want it to be ephemeral (think our results page).

It seems like a getter has 3 caching options, 'always', 'never', and 'default'.  I sort of dislike having two flags to represent 3 options.  What are your thoughts on

```
Getter(['items'], {
  cache: 'always',
})

Getter(['items'], {
  cache: 'default',
})

Getter(['items'], {
  cache: 'never',
})
```

cc: @loganlinn

scjackson commented:
@jordangarcia I thought you were advocating for maintaining the cache by default behavior of `reactor.evaluate` (ie caching by default). One flag seems cleaner to me as well

jordangarcia commented:
ah yeah I was unclear on that.  If we implement an upper limit as well as
an observe cache cleanup then we should be able to keep the same defaults.
I'm not committed to one or the other, it'd be good to build the options in
and then do some real world testing.

On Tue, Jan 26, 2016 at 11:32 AM, Sam Jackson <notifications@github.com>
wrote:

> @jordangarcia <https://github.com/jordangarcia> I thought you were
> advocating for maintaining the cache by default behavior of
> reactor.evaluate (ie caching by default). One flag seems cleaner to me as
> well
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/optimizely/nuclear-js/pull/208#issuecomment-175191735>
> .
>

scjackson commented:
@jordangarcia 

Updated with the following:
1. `Getter` constructor, with `cache` and `cacheKey` options
2. Removing unobserved values from cache
3. Dropped `getCacheValues` and `clearCacheValues` from `Reactor`
4. default max items limit of 1000

scjackson commented:
@jordangarcia 

one other relevant change, I'm now clearing cache values on `Reactor.reset` as well

jordangarcia commented on src/getter.js:
```diff
@@ -9,6 +11,19 @@ import { isKeyPath } from './key-path'
  */
 const identity = (x) => x
 
+class GetterClass {
```
I think this would be better as a pseudo-constructor.  IE I dont like the idea of changing the Getter type because now we have to handle both a `Getter` instance and a plain array getter in every interface.

Instead `Getter(['items'])` returning simply `['items']` makes a lot of sense to me for backwards compat and simplicity.

scjackson replied:
so basically what i had before?

jordangarcia replied:
yes, but called `Getter` instead of `addOptionsToGetter`.  The latter name implied that you could mutate an existing Getter (potentially globally) versus a function to create a Getter.

scjackson replied:
Ok -- `addOptionsToGetter` was by ref. Which would modify it globally. We could return a cloned copy with options added instead. Thoughts?

jordangarcia commented on src/reactor/fns.js:
```diff
@@ -168,6 +169,8 @@ export function loadState(reactorState, state) {
   // use the passed in getter as the key so we can rely on a byreference call for unobserve
   const getterKey = getter
+
+  getter = convertToGetterLiteral(getter)
```
I really dislike how we have to use `convertToGetterLiteral` everywhere.  It's also not a free operation since we do assertion checking that it's infact a valid getter.

scjackson replied:
me too. only did it to support `Getter` type

jordangarcia commented on src/reactor/fns.js:
```diff
@@ -223,65 +226,79 @@ export function getOption(reactorState, option) {
-  return observerState.withMutations(map => {
-    entriesToRemove.forEach(entry => removeObserverByEntry(map, entry))
+  // Update both observer and reactor state
+  observerState = observerState.withMutations(oState => {
```
not a fan of these function params, can we not do `observerState` because of `no-shadow`?

scjackson replied:
Could you clarify? Not sure what you mean by this?

jordangarcia commented on src/reactor/fns.js:
```diff
@@ -223,65 +226,79 @@ export function getOption(reactorState, option) {
+      entriesToRemove.forEach(entry => { removeObserverByEntry(oState, rState, entry) })
+    })
   })
+  return [
```
I think we use objects for multiple return values in other places

```
return {
  observerState,
  reactorState,
}
```

usage:
```
const { observerState, reactorState } = fns.removeObserver(...)
```

loganlinn replied:
:+1:

jordangarcia commented on src/reactor/fns.js:
```diff
@@ -424,6 +445,20 @@ function isCached(reactorState, keyPathOrGetter) {
+
+  // Check cache settings on a getter basis
+  if (isGetterObject(getter) && getter.cache !== 'default') {
+    useCache = getter.cache === 'always' ? true : false
```
```
useCache = getGetterOption(getter, 'cache') === 'always';
```

jordangarcia commented on src/reactor/fns.js:
```diff
@@ -223,65 +226,79 @@ export function getOption(reactorState, option) {
-          return observers
+export function removeObserverByEntry(observerState, reactorState, entry) {
+  return [
+    observerState.withMutations(map => {
```
putting this logic positionally in a return array is complex.

Put in a variable and return both variables.

jordangarcia commented on src/reactor/fns.js:
```diff
@@ -223,65 +226,79 @@ export function getOption(reactorState, option) {
+  // Update both observer and reactor state
+  observerState = observerState.withMutations(oState => {
+    reactorState = reactorState.withMutations(rState => {
+      entriesToRemove.forEach(entry => { removeObserverByEntry(oState, rState, entry) })
```
Are there tests for this behavior?

Also we shouldn't remove the cache entry until we are actually ready to unobserve the getter.   There can be a case where two components are setting up observe on the same entity and we shouldn't kill the cache entry until the getter is no longer being observed.

scjackson replied:
Hmm. I guess we'd have to keep track of the # of observers associated with a given getter. Will look into it.

jordangarcia replied:
I think we already have that logic (the number of observers per getter) in the `observerState` already.

loganlinn commented on src/reactor.js:
```diff
@@ -24,11 +26,24 @@ import {
+    // if user has defined maxItemsToCache, use the number provide or set to null (ie no max)
+    if (config.maxItemsToCache !== undefined) {
+      maxItemsToCache = Number(config.maxItemsToCache)
+      maxItemsToCache = maxItemsToCache && maxItemsToCache > 0 ? maxItemsToCache : null
```
minor: is there a difference between `maxItemsToCache && maxItemsToCache > 0` and just `maxItemsToCache > 0` here? looks like it could be simplified

loganlinn commented on src/reactor/fns.js:
```diff
@@ -434,11 +469,53 @@ function cacheValue(reactorState, getter, value) {
+function updateCacheRecency(reactorState, getter) {
+  const cacheKey = getCacheKey(getter)
+  return reactorState.withMutations(state => {
+    state.deleteIn(['cacheRecency', cacheKey])
```
This `deleteIn()` isn't needed. `setIn()` is called for same key path directly after. 

```js
return reactorState.setIn(['cacheRecency', cacheKey], cacheKey)
```

scjackson commented:
@jordangarcia @loganlinn 

Removed the `Getter` constructor in favor of `Getter` pseudo constructor, modified `removeObserver` to only delete cached getter values if nothing else is observing the getter, or if the user invoked remove by getter rather than by entry.
